File: admin/internal/domain/product/inventory.go
package product

import (
	"context"
	"time"

	"github.com/colinrs/shopjoy/pkg/application"
	"github.com/colinrs/shopjoy/pkg/domain/shared"
	"gorm.io/gorm"
)

// Warehouse represents a physical storage location
type Warehouse struct {
	application.Model
	TenantID  shared.TenantID
	Code      string // 仓库代码
	Name      string // 仓库名称
	Country   string // 所在国家 (ISO country code)
	Address   string // 详细地址
	IsDefault bool   // 是否默认仓库
	Status    shared.Status
	Audit     shared.AuditInfo `gorm:"embedded"`
}

func (w *Warehouse) TableName() string {
	return "warehouses"
}

func (w *Warehouse) Enable() {
	w.Status = shared.StatusEnabled
}

func (w *Warehouse) Disable() {
	w.Status = shared.StatusDisabled
}

func (w *Warehouse) IsEnabled() bool {
	return w.Status == shared.StatusEnabled
}

// WarehouseInventory tracks stock per SKU per warehouse
type WarehouseInventory struct {
	application.Model
	TenantID       shared.TenantID
	SKUCode        string // SKU代码
	WarehouseID    int64
	AvailableStock int              // 可用库存
	LockedStock    int              // 锁定库存
	Audit          shared.AuditInfo `gorm:"embedded"`
}

func (wi *WarehouseInventory) TableName() string {
	return "warehouse_inventories"
}

// TotalStock returns total stock (available + locked)
func (wi *WarehouseInventory) TotalStock() int {
	return wi.AvailableStock + wi.LockedStock
}

// CanDeduct checks if enough stock is available
func (wi *WarehouseInventory) CanDeduct(quantity int) bool {
	return wi.AvailableStock >= quantity
}

// InventoryLog records all stock changes
type InventoryLog struct {
	application.Model
	TenantID       shared.TenantID
	SKUCode        string
	ProductID      int64
	WarehouseID    int64  // 0 = total/summary
	ChangeType     string // manual, order, return, adjustment
	ChangeQuantity int    // positive = increase, negative = decrease
	BeforeStock    int
	AfterStock     int
	OrderNo        string
	Remark         string
	OperatorID     int64
}

func (il *InventoryLog) TableName() string {
	return "inventory_logs"
}

// Inventory change types
const (
	InventoryChangeManual     = "manual"
	InventoryChangeOrder      = "order"
	InventoryChangeReturn     = "return"
	InventoryChangeAdjustment = "adjustment"
)

// WarehouseRepository interface
type WarehouseRepository interface {
	Create(ctx context.Context, db *gorm.DB, warehouse *Warehouse) error
	Update(ctx context.Context, db *gorm.DB, warehouse *Warehouse) error
	Delete(ctx context.Context, db *gorm.DB, id int64) error
	FindByID(ctx context.Context, db *gorm.DB, id int64) (*Warehouse, error)
	FindByCode(ctx context.Context, db *gorm.DB, tenantID int64, code string) (*Warehouse, error)
	FindByTenant(ctx context.Context, db *gorm.DB, tenantID int64) ([]*Warehouse, error)
}

// WarehouseInventoryRepository interface
type WarehouseInventoryRepository interface {
	Create(ctx context.Context, db *gorm.DB, wi *WarehouseInventory) error
	Update(ctx context.Context, db *gorm.DB, wi *WarehouseInventory) error
	FindBySKUAndWarehouse(ctx context.Context, db *gorm.DB, skuCode string, warehouseID int64) (*WarehouseInventory, error)
	FindBySKU(ctx context.Context, db *gorm.DB, skuCode string) ([]*WarehouseInventory, error)
}

// InventoryLogRepository interface
type InventoryLogRepository interface {
	Create(ctx context.Context, db *gorm.DB, log *InventoryLog) error
	Find(ctx context.Context, db *gorm.DB, query InventoryLogQuery) ([]*InventoryLog, int64, error)
}

// InventoryLogQuery for querying inventory logs.
// Optional fields (ProductID, SKUCode, ChangeType) are filter conditions;
// when zero-valued, no filter is applied for that field.
type InventoryLogQuery struct {
	shared.PageQuery
	ProductID  int64
	SKUCode    string
	ChangeType string
	StartTime  time.Time
	EndTime    time.Time
}

// InventoryRepository interface for SKU-level inventory operations
type InventoryRepository interface {
	// SKU inventory
	GetSKUInventory(ctx context.Context, db *gorm.DB, skuCode string) (*SKU, error)
	UpdateSKUStock(ctx context.Context, db *gorm.DB, skuCode string, availableStock, lockedStock int) error
	LockStock(ctx context.Context, db *gorm.DB, skuCode string, quantity int) error
	DeductStock(ctx context.Context, db *gorm.DB, skuCode string, quantity int) error
	RestoreStock(ctx context.Context, db *gorm.DB, skuCode string, quantity int) error

	// Warehouse inventory
	GetWarehouseInventory(ctx context.Context, db *gorm.DB, skuCode string, warehouseID int64) (*WarehouseInventory, error)
	SetWarehouseStock(ctx context.Context, db *gorm.DB, skuCode string, warehouseID int64, stock int) error

	// Logs
	CreateLog(ctx context.Context, db *gorm.DB, log *InventoryLog) error
	GetLogs(ctx context.Context, db *gorm.DB, query InventoryLogQuery) ([]*InventoryLog, int64, error)

	// Low stock
	GetLowStockSKUs(ctx context.Context, db *gorm.DB) ([]*SKU, error)
}

File: shop-admin/src/api/inventory.ts
import request from '@/utils/request'

// Warehouse types
export interface Warehouse {
  id: string
  code: string
  name: string
  country: string
  address: string
  is_default: boolean
  status: number
  created_at: string
  updated_at: string
}

export interface CreateWarehouseRequest {
  code: string
  name: string
  country?: string
  address?: string
  is_default?: boolean
}

export interface UpdateWarehouseRequest {
  id: string
  name: string
  country?: string
  address?: string
  is_default?: boolean
}

// Inventory types
export interface WarehouseInventoryItem {
  warehouse_id: string
  warehouse_name: string
  available_stock: number
  locked_stock: number
}

export interface SKUInventory {
  sku_code: string
  product_id: string
  total_stock: number
  available_stock: number
  locked_stock: number
  safety_stock: number
  is_low_stock: boolean
  warehouses: WarehouseInventoryItem[]
}

export interface InventoryLog {
  id: string
  sku_code: string
  product_id: string
  warehouse_id: string
  change_type: string
  change_quantity: number
  before_stock: number
  after_stock: number
  order_no: string
  remark: string
  operator_id: string
  created_at: string
}

export interface LowStockSKU {
  sku_code: string
  product_id: string
  product_name: string
  available_stock: number
  safety_stock: number
}

// Warehouse API functions
export function getWarehouses() {
  return request<{ list: Warehouse[] }>({
    url: '/api/v1/warehouses',
    method: 'get'
  })
}

export function getWarehouse(id: string) {
  return request<Warehouse>({
    url: `/api/v1/warehouses/${id}`,
    method: 'get'
  })
}

export function createWarehouse(data: CreateWarehouseRequest) {
  return request<{ id: string }>({
    url: '/api/v1/warehouses',
    method: 'post',
    data
  })
}

export function updateWarehouse(data: UpdateWarehouseRequest) {
  return request<Warehouse>({
    url: `/api/v1/warehouses/${data.id}`,
    method: 'put',
    data
  })
}

export function updateWarehouseStatus(id: string, status: number) {
  return request<Warehouse>({
    url: `/api/v1/warehouses/${id}/status`,
    method: 'put',
    data: { id, status }
  })
}

export function setDefaultWarehouse(id: string) {
  return request<Warehouse>({
    url: `/api/v1/warehouses/${id}/default`,
    method: 'put'
  })
}

export function deleteWarehouse(id: string) {
  return request<{ id: string }>({
    url: `/api/v1/warehouses/${id}`,
    method: 'delete'
  })
}

// Inventory API functions
export function getSKUInventory(skuCode: string) {
  return request<SKUInventory>({
    url: `/api/v1/inventory/sku/${skuCode}`,
    method: 'get'
  })
}

export function updateSKUStock(data: {
  sku_code: string
  warehouse_id?: string
  available_stock: number
  remark?: string
}) {
  return request<{ id: string }>({
    url: '/api/v1/inventory/stock',
    method: 'put',
    data
  })
}

export function adjustStock(data: {
  sku_code: string
  warehouse_id: string
  quantity: number
  remark?: string
}) {
  return request<{ id: string }>({
    url: '/api/v1/inventory/adjust',
    method: 'post',
    data
  })
}

export function getInventoryLogs(params?: {
  page?: number
  page_size?: number
  sku_code?: string
  product_id?: string
  type?: string
}) {
  return request<{ list: InventoryLog[]; total: number }>({
    url: '/api/v1/inventory/logs',
    method: 'get',
    params
  })
}

export interface ExportInventoryLogsParams {
  sku_code?: string
  product_id?: string
  type?: string
  [key: string]: unknown
}

/**
 * Export inventory logs - returns URL and params for download utility
 */
export function exportInventoryLogsUrl(params: ExportInventoryLogsParams): { url: string; params: ExportInventoryLogsParams } {
  return {
    url: '/api/v1/inventory/logs/export',
    params
  }
}

export function getLowStockSKUs(params?: { page?: number; page_size?: number }) {
  return request<{ list: LowStockSKU[]; total: number }>({
    url: '/api/v1/inventory/low-stock',
    method: 'get',
    params
  })
}

export function batchUpdateSafetyStock(items: { sku_code: string; safety_stock: number }[]) {
  return request<{ id: string }>({
    url: '/api/v1/inventory/safety-stock',
    method: 'put',
    data: { items }
  })
}